scene/2d/physics_body_2d.cpp

```diff
@@ -1049,134 +1052,234 @@ void RigidBody2D::_reload_physics_characteristics() {
+					if (was_on_floor && !on_floor && result.motion.dot(up_direction) > 0 && linear_velocity.dot(up_direction) <= 0) {
 						Transform2D gt = get_global_transform();
-						if (result.motion.length() > margin) {
-							gt.elements[2] -= result.motion.slide(up_direction);
-						} else {
-							gt.elements[2] -= result.motion;
-						}
+						gt.elements[2] -= up_direction * up_direction.dot(result.motion);
 						set_global_transform(gt);
+						on_floor = true;
+						platform_rid = prev_platform_rid;
+						platform_layer = prev_platform_layer;
+
+						platform_velocity = prev_platform_velocity;
+						floor_normal = prev_floor_normal;
 						linear_velocity = Vector2();
-						return;
+						motion = Vector2();
+						break;
+					}
+					// Prevent to move against wall on air.
+					else if (sliding_enabled && !on_floor) {
+						motion = up_direction * up_direction.dot(result.remainder);
+						motion = motion.slide(result.collision_normal);
+					} else {
+						motion = result.remainder;
 					}
 				}
-
-				if (sliding_enabled || !on_floor) {
-					motion = result.remainder.slide(result.collision_normal);
-					linear_velocity = linear_velocity.slide(result.collision_normal);
+				// Constant Speed when the slope is upward.
+				else if (constant_speed_on_floor && is_on_floor_only() && can_apply_constant_speed && was_on_floor && motion.dot(result.collision_normal) < 0) {
+					can_apply_constant_speed = false;
+					Vector2 motion_slide_norm = result.remainder.slide(result.collision_normal).normalized();
+					if (!motion_slide_norm.is_equal_approx(Vector2())) {
+						motion = motion_slide_norm * (motion_slide_up.length() - result.motion.slide(up_direction).length() - last_travel.slide(up_direction).length());
+					}
+				} else if ((sliding_enabled || !on_floor) && !(on_ceiling && !slide_on_ceiling && linear_velocity.dot(up_direction) > 0)) {
+					Vector2 slide_motion = result.remainder.slide(result.collision_normal);
+					if (slide_motion.dot(linear_velocity) > 0.0) {
+						motion = slide_motion;
+					} else {
+						motion = Vector2();
+					}
+					if (slide_on_ceiling && on_ceiling) {
+						if (linear_velocity.dot(up_direction) > 0) {
+							linear_velocity = linear_velocity.slide(result.collision_normal);
+						} else {
+							// Avoid acceleration in slope when falling.
+							linear_velocity = up_direction * up_direction.dot(linear_velocity);
+						}
+					}
 				} else {
 					motion = result.remainder;
+					if (on_ceiling && !slide_on_ceiling && linear_velocity.dot(up_direction) > 0) {
+						linear_velocity = linear_velocity.slide(up_direction);
+						motion = motion.slide(up_direction);
+					}
 				}
-			}
 
-			sliding_enabled = true;
+				last_travel = result.motion;
+			}
+			// Apply constant speed when the slope is downward.
+			else if (i == 0 && constant_speed_on_floor && first_slide && was_on_floor && !Math::is_equal_approx(floor_snap_length, 0) && _on_floor_if_snapped(was_on_floor)) {
```

Comment:
And a bit more details here in comments to explain what this is doing (what you explained to me in the other review comment).